### PR TITLE
Tree: Initial implementation of change mark lineage

### PR DESCRIPTION
**Pull request by alex-pardes**
Records lineage of attaches when rebasing over detaches, and uses lineage comparisons to preserve the relative order of concurrent attaches even when their target positions are no longer distinct.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 1]
```diff
@@ -146,6 +165,78 @@ function rebaseMark<TNodeChange>(
...
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+function isRelevantToLineage<TNodeChange>(mark: Mark<TNodeChange>): boolean {
+    // TODO
+    return !isSkipMark(mark) && mark.type === "Delete";
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:225]: I presume we'd want "MDelete" to also be considered relevant to lineage.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 2]
```diff
@@ -146,6 +165,78 @@ function rebaseMark<TNodeChange>(
...
+            } else if (offset1 > offset2) {
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+function isRelevantToLineage<TNodeChange>(mark: Mark<TNodeChange>): boolean {
+    // TODO
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:224]: Is this TODO meant to be dealt with before merging the PR?

File: packages/dds/tree/src/feature-libraries/sequence-change-family/changeset/format.ts [hunk 3]
```diff
@@ -107,6 +107,16 @@ export namespace Transposed {
...
+
+        /**
+         * Record of relevant information about changes this mark has been rebased over.
+         */
+        lineage?: LineageEvent[];
+    }
+
+    export interface LineageEvent {
+        readonly revision: RevisionTag;
+        readonly offset: number;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-change-family/changeset/format.ts:119]: The sequence-change-family changeset is mostly dead code being kept because it has code and test cases that have yet to be migrated to the sequence-field implementation and tests. Feel free to drop the changes to this file.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 4]
```diff
@@ -48,57 +48,60 @@ export type NodeChangeRebaser<TNodeChange> = (
...
+        } else if (currMark === undefined) {
+            assert(
+                baseMark !== undefined,
+                "Loop condition should prevent both iterators from being empty",
+            );
+            baseIter.pop();
+            if (isRelevantToLineage(baseMark)) {
+                baseDetachOffset += getInputLength(baseMark);
+            } else {
+                break;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:83]: Don't we need to set `baseDetachOffset` to zero here?

Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:83]: Are we depending on it not being set to zero at line 135?

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 5]
```diff
@@ -146,6 +165,78 @@ function rebaseMark<TNodeChange>(
...
+
+function handleCurrAttach<T>(
+    currMark: Attach<T>,
+    factory: MarkListFactory<T>,
+    lineageRequests: LineageRequest<T>[],
+    offset: number,
+) {
+    const rebasedMark = clone(currMark);
+    factory.pushContent(rebasedMark);
+    lineageRequests.push({ mark: rebasedMark, offset });
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:180]: Why do we need to defer the construction of the lineage information here. IOW, why not have this function take the base `RevisionTag` and just do `mark.lineage.push({ revision, offset: offset });` (modulo the undefined `mark.lineage` array case)?

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts [hunk 6]
```diff
@@ -4,8 +4,12 @@
  */
 
 import { strict as assert } from "assert";
+import { Delta } from "../../../core";
 import { SequenceField as SF } from "../../../feature-libraries";
-import { makeAnonChange } from "../../../rebase";
+
+// eslint-disable-next-line import/no-internal-modules
+import { MarkListFactory } from "../../../feature-libraries/sequence-field";
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:11]: Use `SF.MarkListFactory` instead.

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts [hunk 7]
```diff
@@ -352,4 +377,62 @@ describe("SequenceField - Rebase", () => {
...
+        const insertE = tagChange(createInsertChangeset(3, 1), brand(3));
+        const insertD2 = rebaseTagged(insertD, delA, delB, delC);
+        const insertE2 = rebaseTagged(insertE, delA, delB, delC, insertD2);
+        const expected = createInsertChangeset(1, 1);
+        assertDeltaEqual(insertE2.change, expected);
+    });
 });
+
+function assertDeltaEqual(actual: TestChangeset, expected: TestChangeset) {
+    assert.deepEqual(toDelta(actual), toDelta(expected));
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:406]: Deep equality on delta will fail in some cases when it shouldn't. This is not a problem here if the tests are passing but we should avoid using it to prevent it from propagating though copy-paste. Use `assertDeltaEqual` from `src\test\utils.ts` instead.

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts [hunk 8]
```diff
@@ -352,4 +377,62 @@ describe("SequenceField - Rebase", () => {
         ];
         assert.deepEqual(actual, expected);
     });
+
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:380]: It would be nice to have test cases for:
- A pair of inserts targetting interior gaps of a lengthy delete
- A pair of inserts where the later insert is meant to land before the earlier insert

File: packages/dds/tree/src/feature-libraries/sequence-field/format.ts [hunk 9]
```diff
@@ -105,12 +119,13 @@ export interface ModifyDetach<TNodeChange = NodeChangeType> extends HasOpId {
     changes: TNodeChange;
 }
 
-export interface Reattach extends HasOpId {
+// TODO: Reattach and ModifyReattach shouldn't have tiebreak field
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/format.ts:122]: Feel free to split the tiebreak off from the HasPlaceFields or however you'd want it refactored.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 10]
```diff
@@ -146,6 +165,78 @@ function rebaseMark<TNodeChange>(
...
+    // TODO: Handle tiebreak
+    return false;
+}
+
+function compareLineages(
+    lineage1: LineageEvent[] | undefined,
+    lineage2: LineageEvent[] | undefined,
+): number {
+    if (lineage1 === undefined || lineage2 === undefined) {
+        return 0;
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:200]: This doesn't seem to allow for cases where we wan to return "they need to be tie-breaked by according to the base mark's tiebreak".
Here's the example scenario we discussed before:
Staring state: a b
A1) Delete a b
A2) Insert c
B) Insert x between a and b

Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:200]: Also, it would be nice to add a test for this scenario.

File: packages/dds/tree/src/feature-libraries/sequence-field/format.ts [hunk 11]
```diff
@@ -59,6 +57,20 @@ export interface HasPlaceFields {
...
+     * Record of relevant information about changes this mark has been rebased over.
+     */
+    lineage?: LineageEvent[];
+}
+
+export interface LineageEvent {
+    readonly revision: RevisionTag;
+
+    /**
+     * The position of this mark within the range of detached nodes in this revision
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/format.ts:71]: Is that acrosss all detached nodes for that field for that revision, or only the detached nodes that merged into the gap that the mark sporting this lienage targets? Would be good to clarify in the comment.

File: packages/dds/tree/src/feature-libraries/sequence-field/format.ts [hunk 12]
```diff
@@ -59,6 +57,20 @@ export interface HasPlaceFields {
      * Omit if `Tiebreak.Right` for terseness.
      */
     tiebreak?: Tiebreak;
+
+    /**
+     * Record of relevant information about changes this mark has been rebased over.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/format.ts:62]: Would be good to clarify how those are ordered.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 5 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:180]: Originally I thought we might need to capture information about later base marks. That didn't end up being necessary, but I think we may need to do that in the future. Currently the only reason for doing it this way is that when processing this attach, we do not yet know if the base change has a detach immediately following this attach's target position. If the offset is zero and there is no detach following this attach, we do not want to add a lineage event.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 13]
```diff
@@ -146,6 +165,78 @@ function rebaseMark<TNodeChange>(
...
+    }
+
+    const lineage1Offsets = new Map<RevisionTag, number>();
+    for (const event of lineage1) {
+        lineage1Offsets.set(event.revision, event.offset);
+    }
+
+    for (let i = lineage2.length - 1; i >= 0; i--) {
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:208]: It would be nice to add a comment explaining that, while two marks could bear unrelated lineage information for a given revision (let's call it "rev Foo") (meaning each mark was affected by independent detaches at that revision), the fact that we're walking through the lineage from most recent to most ancient lineage nodes ensures we won't ever compare the indices of such nodes because in order for two marks to end up in the same gap despite having such unrelated lineage nodes from the same revision (rev Foo), they would have needed to be affected by a deletion that merged the independent gaps that they were targeting, and we'll encounter the lineage nodes for that revision before we encounter the lineage nodes for rev Foo.
Hopefully you can find a cleaner way to say all that, but I think it's not obvious from the code that it works at it should because of these facts.

Also, it would be nice to add a test with such a rev Foo.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 10 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:200]: This function returns zero when the lineage does not determine the mark order. In that case we can use the tiebreak policy in `isAttachAfterBaseAttach`.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 5 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:180]: If I'm following correctly, your last sentence implies that we would want to add a lineage event if the offset were zero but there is a detach (in the base change) following this attach. Am I understanding correctly that this is because we want to be able to differentiate an insert at the left edge of a deleted region that used to target the gap before the deleted node from an insert that just targets the "crater" gap left by that same deletion?

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:180]: Yes

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 10 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:201]: I agree there's a code-structure question here. My point is that if this function just returns 0 in the scenario above then it's on `isAttachAfterBaseAttach` to go dig in the lineage to figure out whether the tiebreak of the base insert or the tiebreak from the curr insert should be obeyed. I'm open to either code-structure.

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:201]: Ah, I forgot our earlier conclusion about the correct behavior of this scenario. Thanks for reminding me.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 5 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:181]: A small comment explaining that would be nice.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 4 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:84]: Ok, it looks like we're just saving some lines of code by relying on the updateLineage that's after the loop. Closing this.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts [hunk 14]
```diff
@@ -48,57 +49,60 @@ export type NodeChangeRebaser<TNodeChange> = (
...
+                // TODO: Handle tiebreaking is baseMark is also an attach.
+                currIter.pop();
+                handleCurrAttach(currMark, factory, lineageRequests, baseDetachOffset);
+            }
         } else if (isAttach(baseMark)) {
-            // We currently ignore the ways in which curr marks overlap with these attaches.
-            // These are:
-            // 1. Slice ranges that include prior insertions
-            // We ignore #1 because we do not yet support slices.
+            baseIter.pop();
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:96]: Do we not need to do this stuff here:
```
if (baseDetachOffset > 0 && baseRevision !== undefined) {
    updateLineage(lineageRequests, baseRevision);
}

lineageRequests.length = 0;
baseDetachOffset = 0;
```

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:96]: I don't see why it would be necessary. In fact, it would be incorrect to do that here. In the scheme implemented in this PR, we can only record one lineage event per base changeset. A lineage event currently represents a detach of a contiguous range of nodes. The base changeset could have a detach, immediately followed by an attach, immediately followed by another detach. In that case a contiguous range of nodes is being detached by the two detaches, and so we must record it as a single lineage event.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 10 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:201]: I don't think it requires a significant restructuring of the code, but I would prefer to implement this in a separate PR.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 14 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:96]: Ah, I understand better now. It's probably painfully obvious at that point but I find the logic really hard to follow (and therefore hard to maintain in the future). I don't have a suggestion for making it clearer but I'll think about it more.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 10 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:201]: Sounds good. Please add a TODO comment in case we don't get to it immediately.

File: packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts (hunk 14 above)
Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:96]: I'd rather not spend too much time on this for this PR. We will have to make changes to the lineage representation to support slice moves and also to enable marks to share a lineage graph, and I suspect the algorithm used here will also have to be changed to accommodate.

Comment (yann-achard-MS) [reply packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts:96]: Those future changes may not happen for months, and when they do, I'd rather the code being changed be simpler so we can more easily think about the changes. That said, if you're satisfied with it as is and I don't have something better to suggest, then I wouldn't want hold the PR up either. Resolving this thread. I'll pipe up if I do have something to suggest.

File: packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts (hunk 8 above)
Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/sequence-field/rebase.spec.ts:377]: But those could be tackled in later PRs. Feel free to resolve this comment.

**Review by yann-achard-MS — APPROVED** [review]
